File: 530_Minimum_Absolute_Difference_in_BST.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/minimum-absolute-difference-in-bst/
# Given a binary search tree with non-negative values, find the minimum absolute difference between values of
# any two nodes.

# Inorder traversal, visits nodes in order of ascending value. Track the value of the previous node visited. For each
# node update the min_diff with the difference between node.val and prev, if lower than current min_diff.
# Time - O(n)
# Space - O(1)

class Solution(object):
    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.min_diff = float("inf")
        self.prev = float("-inf")

        def inorder(node):

            if not node:
                return

            inorder(node.left)

            self.min_diff = min(self.min_diff, node.val - self.prev)
            self.prev = node.val

            inorder(node.right)

        inorder(root)
        return self.min_diff

File: java/530_Minimum_Absolute_Difference_in_BST.java
/*
https://leetcode.com/problems/minimum-absolute-difference-in-bst/
Given a binary search tree with non-negative values, find the minimum absolute difference between values of any
two nodes.

Inorder traversal. Update minDiff if not first node. Update previous.
Time - O(n)
Space - O(i)
*/

class Solution {
    private int previous;
    private int minDiff = -1;   // -1 no nodes visited yet

    private void inOrder(TreeNode node) {
        if (node == null)
            return;

        inOrder(node.left);

        if (minDiff != -1)      // not first node, update max
            minDiff = Math.min(minDiff, Math.abs(node.val - previous));
        else                    // first node
            minDiff = Integer.MAX_VALUE;
        previous = node.val;

        inOrder(node.right);
    }

    public int getMinimumDifference(TreeNode root) {
        inOrder(root);
        return minDiff;
    }
}
